### PR TITLE
feat(storage): filecache lazy load, unit-level refill, reduce insert latency

### LINKED ISSUE #12656: bug: compute node recovery fails when file cache is large on cloud
### Describe the bug

RisingWave may crash due to temporary network failure and recover later. 

On cloud, k8s is responsible for health checks. (e.g. On RisingWave inner test cloud setup, the frequency of health check is 10s) If health checks fail several times, the failure pod will be re-deployed.

However, with a large file cache, the recovery of RisingWave can be slow. The compute node may be continuously killed by k8s health check.

e.g.

In this case, compute nodes keep restarting but cannot recover after a failure, but there is no unexpected log between launchings.

image

With the k8s event log, we can see that the compute node is killed by health check:

```
LAST SEEN            TYPE     REASON    OBJECT                                 MESSAGE
53m (x3 over 113m)   Normal   Created   Pod/benchmark-risingwave-compute-c-0   Created container compute
53m (x3 over 113m)   Normal   Started   Pod/benchmark-risingwave-compute-c-0   Started container compute
29m (x9 over 54m)    Normal   Pulled    Pod/benchmark-risingwave-compute-c-0   Container image "ghcr.io/risingwavelabs/risingwave:git-52483b367f65b9f4d8224567fe15f4d2a9407232" already present on machine
4m50s (x174 over 54m)   Warning   Unhealthy   Pod/benchmark-risingwave-compute-c-0   Startup probe failed: dial tcp 10.0.63.110:5688: connect: connection refused
53m                     Normal    Killing     Pod/benchmark-risingwave-compute-c-0   Container compute failed startup probe, will be restarted
9m45s (x125 over 47m)   Warning   BackOff     Pod/benchmark-risingwave-compute-c-0   Back-off restarting failed container compute in pod benchmark-risingwave-compute-c-0_benchmark-yao-manual(0cf79b54-eed1-4185-a5a6-3db8e9fde3b8)
0s (x161 over 47m)      Warning   BackOff     Pod/benchmark-risingwave-compute-c-0   Back-off restarting failed container compute in pod benchmark-risingwave-compute-c-0_benchmark-yao-manual(0cf79b54-eed1-4185-a5a6-3db8e9fde3b8)
0s (x215 over 59m)      Warning   Unhealthy   Pod/benchmark-risingwave-compute-c-0   Startup probe failed: dial tcp 10.0.63.110:5688: connect: connection refused
0s (x195 over 57m)      Warning   BackOff     Pod/benchmark-risingwave-compute-c-0   Back-off restarting failed container compute in pod benchmark-risingwave-compute-c-0_benchmark-yao-manual(0cf79b54-eed1-4185-a5a6-3db8e9fde3b8)
```

The health check failed to connect to the serving port of the compute node.

But the log shows that the compute node served normally at least for a short period. 

image

To make sure the port can be correctly connected, I manually checked it:

image

But because of the slow file cache recovery, the compute node missed some health checks and was killed seconds after it was finally on.

image

Solution:

Lazy load file cache.

The file cache should be created once it is opened, and its recovery can be scheduled in the background. During the file cache recovery, any operation to the file cache acts like an operation to a 0-capacity file cache. Once the file cache is fully recovered, it will serve normally.

File cache lazy loading can be useful for all file cache users. So it will be implemented in the file cache engine (foyer). And that will also keep RisingWave simple.

### Error message/log

_No response_

### To Reproduce

_No response_

### Expected behavior

_No response_

### How did you deploy RisingWave?

_No response_

### The version of RisingWave

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by MrCroxx**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR includes both foyer updates and file cache updates in RisingWave.

Notable updates in foyer part (`41b1d39` ~ `f097abe`):
- fair allocation order
- sharded allocators
- open lazy load
- merge concurrent inflight physical reads
- integrate separate runtime
- interface refactors

Changes in RisingWave part:
- use foyer with runtime + lazy load features
- unit-level data file cache refill
- expose more configurations about cache refill
- increase foyer buffer pool for better fit
- refactors

Close #12656.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by wangrunji0408 — APPROVED** [review]
LGTM for Cargo.lock.

File: src/storage/src/hummock/event_handler/refiller.rs [hunk 1]
```diff
@@ -294,29 +310,107 @@ impl CacheRefillTask {
...
+            };
+            tasks.push(task);
+        }
 
-                let permit = context.concurrency.acquire().await.unwrap();
+        join_all(tasks).await;
+    }
 
-                GLOBAL_CACHE_REFILL_METRICS.data_refill_started_total.inc();
+    async fn data_file_cache_refill_impl(
```
Comment (Li0k) [inline src/storage/src/hummock/event_handler/refiller.rs:323]: How about add some comments for the new refill strategy with unit-level ?

**Review by Li0k — APPROVED** [review]
Rest LGTM, thanks for the PR

File: src/storage/src/hummock/event_handler/refiller.rs (hunk 1 above)
Comment (MrCroxx) [reply src/storage/src/hummock/event_handler/refiller.rs:346]: There are comments on the cache refill config about unit-level cache refill. But I agree with you, more details are needed here.

File: src/common/src/config.rs [hunk 2]
```diff
@@ -595,6 +595,18 @@ pub struct CacheRefillConfig {
     #[serde(default = "default::cache_refill::concurrency")]
     pub concurrency: usize,
 
+    #[serde(default = "default::cache_refill::unit")]
+    pub unit: usize,
```
Comment (hzxa21) [inline src/common/src/config.rs:599]: It is better to add some comments about this config since its name is not very instructive like other configs.

**Review by hzxa21 — APPROVED** [review]
Generally LGTM.

I need to catch up with the foyer changes to better understand the detail changes of this PR. The framework looks good overall so feel free to merge this PR if you think it is ready.

**Pull request merged**